From b4e016f65faf1b458ac659286863bd6664b5a1a7 Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 11:38:50 -0400
Subject: [ethernet] i40e: Allow RSS table entry range and GPS to be any number, not necessarily power of 2

Message-id: <1409658042-22689-170-git-send-email-sassmann@redhat.com>
Patchwork-id: 90361
O-Subject: [RHEL7.1 PATCH 169/281] i40e: Allow RSS table entry range and GPS to be any number, not necessarily power of 2
Bugzilla: 1091126
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Anjali Singhai Jain <anjali.singhai@intel.com>

We tell the HW upper boundary of power of 2 in VSI config,
but the HW does not restrict us to use just power of 2 GPS in
case of RSS as long as we are not sharing the RSS table with
another VSI (VMDq). We at present are not doing RSS in VMDq
VSI.
If we were to enable that and if the system had CPU count which
was not power 2, the VMDq VSIs will see a little skewed distribution.

Change-ID: I3ea797ce9065a3ca4fc4d04251bf195463410473
Signed-off-by: Anjali Singhai Jain <anjali.singhai@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit eb051afedf0ace9bf1356d5178f08ef00e6e3249)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40e/i40e_main.c b/drivers/net/ethernet/intel/i40e/i40e_main.c
index 2aa212e..009debf 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_main.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_main.c
@@ -1366,7 +1366,7 @@ static void i40e_vsi_setup_queue_map(struct i40e_vsi *vsi,
  vsi->tc_config.numtc = numtc;
  vsi->tc_config.enabled_tc = enabled_tc ? enabled_tc : 1;
  /* Number of queues per enabled TC */
- num_tc_qps = rounddown_pow_of_two(vsi->alloc_queue_pairs/numtc);
+ num_tc_qps = vsi->alloc_queue_pairs/numtc;
  num_tc_qps = min_t(int, num_tc_qps, I40E_MAX_QUEUES_PER_TC);
 
  /* Setup queue offset/count for all TCs for given VSI */
@@ -6498,7 +6498,6 @@ int i40e_reconfig_rss_queues(struct i40e_pf *pf, int queue_count)
   return 0;
 
  queue_count = min_t(int, queue_count, pf->rss_size_max);
- queue_count = rounddown_pow_of_two(queue_count);
 
  if (queue_count != pf->rss_size) {
   i40e_prep_for_reset(pf);
@@ -6554,7 +6553,6 @@ static int i40e_sw_init(struct i40e_pf *pf)
  if (pf->hw.func_caps.rss) {
   pf->flags |= I40E_FLAG_RSS_ENABLED;
   pf->rss_size = min_t(int, pf->rss_size_max, num_online_cpus());
-  pf->rss_size = rounddown_pow_of_two(pf->rss_size);
  } else {
   pf->rss_size = 1;
  }
-- 
1.7.1